### PR TITLE
Add issues collection step

### PR DESCRIPTION
**Pull request by RobiNino**

### CODE REVIEW

File: src/main/java/org/jfrog/hudson/pipeline/common/types/buildInfo/BuildInfo.java [hunk 1]
```diff
@@ -63,11 +64,13 @@ public BuildInfo(Run build) {
         this();
         this.name = BuildUniqueIdentifierHelper.getBuildName(build);
         this.number = BuildUniqueIdentifierHelper.getBuildNumber(build);
+        issues.setBuildName(name);
```
Comment (yahavi) [inline src/main/java/org/jfrog/hudson/pipeline/common/types/buildInfo/BuildInfo.java]: Please add for consistency:
```suggestion
        this.issues.setBuildName(name);
```

File: src/main/java/org/jfrog/hudson/pipeline/common/types/buildInfo/BuildInfo.java [hunk 2]
```diff
@@ -63,11 +64,13 @@ public BuildInfo(Run build) {
...
     }
 
     @Whitelisted
     public void setName(String name) {
         this.name = name;
+        issues.setBuildName(name);
```
Comment (yahavi) [inline src/main/java/org/jfrog/hudson/pipeline/common/types/buildInfo/BuildInfo.java]: Please add for consistency:
```suggestion
        this.issues.setBuildName(name);
```

File: src/main/java/org/jfrog/hudson/pipeline/common/BuildInfoDeployer.java [hunk 3]
```diff
@@ -56,20 +57,39 @@ public BuildInfoDeployer(ArtifactoryConfigurator configurator, ArtifactoryBuildI
...
+        }
+
+        addVcsDataToBuild(build, buildinfoAccessor);
+    }
+
+    private void addVcsDataToBuild(Run build, BuildInfoAccessor buildinfoAccessor) {
+        List<Vcs> vcsList = getVcsFromGitPlugin(build);
+
+        // If collected VCS in a different flow
+        if (buildinfoAccessor.getVcs() != null && !buildinfoAccessor.getVcs().isEmpty()) {
```
Comment (yahavi) [inline src/main/java/org/jfrog/hudson/pipeline/common/BuildInfoDeployer.java]: I think that this 'if' is redundant

File: src/main/java/org/jfrog/hudson/pipeline/scripted/steps/CollectIssuesStep.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
+package org.jfrog.hudson.pipeline.scripted.steps;
```
Comment (yahavi) [inline src/main/java/org/jfrog/hudson/pipeline/scripted/steps/CollectIssuesStep.java]: This file has many warnings. Please fix them!

File: src/main/java/org/jfrog/hudson/pipeline/declarative/steps/CollectIssuesStep.java [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+        @Override
+        protected Void run() throws Exception {
+            // Set spec
+            SpecConfiguration specConfiguration = new SpecConfiguration(step.config, step.configPath);
+            config = SpecUtils.getSpecStringFromSpecConf(specConfiguration, env, ws, listener.getLogger());
+
+            // Get build info
+            BuildInfo buildInfo = DeclarativePipelineUtils.getBuildInfo(ws, build, step.customBuildName, step.customBuildNumber);
+
+            buildInfo.getIssues().collectBuildIssues(build, listener, ws, buildInfo.getName(), config);
```
Comment (yahavi) [inline src/main/java/org/jfrog/hudson/pipeline/declarative/steps/CollectIssuesStep.java]: Can you implement the collectBuildIssues as an executor?

**Comment by reenberg** [comment]
Is there any documentation on this?  

I assume that this solves the comment on the JFrog [Jenkins Artifactory Plug-in / JIRA Integration](https://www.jfrog.com/confluence/display/RTF/Jenkins+Artifactory+Plug-in#JenkinsArtifactoryPlug-in-JIRAIntegration)?:  

> JIRA Integration is currently supported only in Free-Style and Maven jobs.

In which case the above mentioned wiki page is outdated/wrong.

**Comment by eyalbe4** [comment]
@reenberg,
The new Issues Collection API is available for Pipeline Jobs. See the documentation for the Declarative and Scripted syntax [here](https://www.jfrog.com/confluence/display/RTF/Working+With+Pipeline+Jobs+in+Jenkins). You can also see the github examples referenced from these documentation pages.
The old JIRA integration is still there, but it is supported only for Free-Style jobs.
Please let me know if you need additional information or help.

**Comment by reenberg** [comment]
Nice. Seems to explain it quite fine.  I must have just missed the changes you added :)

It seems that this addition only caters for issue IDs in commit messages.  Is there any plans on supporting issue IDs from the branch name as well?

Most workflows I work with, that includes JIRA, starts with a branch being created in bitbucket from a pre-existing JIRA ticket,.  This way JIRA creates the branch with the information from the issue summary, something like "feature/HAP-1231-add-issues-collection-step-to-pipeline".   And as most branches are only created to solve one issue, the issue ID is never added to the commit ID, as it is in the branch name.

However I guess, that it will work, when the branch is merged, as the merge commit will contain the branch name (assuming one doesn't use fast forward only).

File: src/main/java/org/jfrog/hudson/pipeline/common/executors/PublishBuildInfoExecutor.java [hunk 6]
```diff
@@ -14,17 +15,20 @@
...
         this.buildInfo = buildInfo;
         this.listener = listener;
         this.build = build;
+        this.ws = ws;
     }
 
     @Override
     public void execute() throws Exception {
         BuildInfoAccessor buildInfoAccessor = new BuildInfoAccessor(buildInfo);
+        buildInfoAccessor.appendVcs(Utils.extractVcs(ws));
```
Comment (casz) [inline src/main/java/org/jfrog/hudson/pipeline/common/executors/PublishBuildInfoExecutor.java]: You already have access to the `Run` where you can access the build's ENV

Somewhat 😢 that publishBuildInfo requires an agent to publish this step.

```java
EnvVars env = build.getEnvironment(listener)
env.get("GIT_URL")
```

You should be able to loop over `GIT_URL_1`, `GIT_URL_2`

See

https://github.com/jenkinsci/git-plugin/blob/2e6d1bf0d10e04632f8153a3d2d9fc4d7d7c8bf2/src/main/java/hudson/plugins/git/GitSCM.java#L1368-L1376

**Pull request merged**